pytorch_lightning/metrics/functional/classification.py

```diff
@@ -872,6 +873,89 @@ def average_precision(
     return -torch.sum((recall[1:] - recall[:-1]) * precision[:-1])
 
 
+def object_detection_mean_average_precision(
+    pred: torch.Tensor, target: torch.Tensor, iou_threshold: float, num_classes: int
+) -> torch.Tensor:
+    """
+    Compute mean average precision for object detection task. Calculates average precision
+    using AUC.
+
+    Args:
+        pred: Tensor containing predictions, where each row is of the following format:
+              [image_idx, class_pred, class_prob, x_min, y_min, x_max, y_max]
+        target: Tensor containing ground truths, where each row is of the follwing format:
+                [image_idx, class_label, x_min, y_min, x_max, y_max]
+        iou_threshold: threshold for IoU score for determining true positive and
+                       false positive predictions.
+        num_classes: number of total classes
+
+    Returns:
+        mean of the average precision for each class in object detection task.
+    """
+    average_precisions = torch.zeros(num_classes)
+    for c in range(num_classes):
+        c_pred = [p for p in pred if p[1] == c]
+        c_target = [t for t in target if t[1] == c]
+        if len(c_target) == 0:
+            continue
+        # Sort c_pred in descending order w.r.t. confidence score
+        c_pred = sorted(c_pred, key=lambda x: x[2], reverse=True)
+        num_targets_per_image = Counter(t[0].item() for t in c_target)
+        targets_tracker = {
+            image_idx: torch.zeros(count)
+            for image_idx, count in num_targets_per_image.items()
+        }
+        tps, fps = torch.zeros(len(c_pred)), torch.zeros(len(c_pred))
+        for i, p in enumerate(c_pred):
+            # Get the targets that correspond to the same image as the prediction
+            ground_truths = [t for t in c_target if t[0] == p[0]]
+            best_iou = 0
+            best_target_idx = 0
+            for j, t in enumerate(ground_truths):
+                curr_iou = object_detection_iou(p[None, 3:], t[None, 2:])
+                if curr_iou > best_iou:
+                    best_iou = curr_iou
+                    best_target_idx = j
+            image_idx = int(p[0].item())
+            if (
+                best_iou > iou_threshold
```

Comment:
Do you have a pointer to the standard mAP implementation? The sources that I looked are not too reliable, but my understanding is that you should call it a TP if IoU > threshold with any ground truth box that hasn't been assigned to another predicted box already. For example, the `get_mappings()` function in this [Kaggle notebook](https://www.kaggle.com/kshitijpatil09/pytorch-mean-absolute-precision-calculation) (by @Kshitij09 ) first filters out the ground truth boxes that have already been assigned and then finds the max IoU.

Reply:
If you check out some of the links referenced above, you'll see some implementation references. @LucaButera is right in that the standard mAP implementation compares only with the best match.

Reply:
The [implementation](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/Evaluator.py#L104) from @rafaelpadilla works identically to your implementation.

The implementation from @tarlen5 uses a different logic. [Every occurrence of IoU > threshold is saved in a list.](https://gist.github.com/tarlen5/008809c3decf19313de216b9208f3734#file-calculate_mean_ap-py-L114) Then the occurrences are iterated starting from the highest IoU and [if both the predicted and the ground truth box don't have a match yet, they are added to the matches](https://gist.github.com/tarlen5/008809c3decf19313de216b9208f3734#file-calculate_mean_ap-py-L132).

The latter implementation is able to match the boxes regardless of whether it's the best match or not. And it also has the nice property that it starts from the pair of boxes with the highest IoU. The algorithm in this pull request is greedy in the sense that we might match a ground truth to the first predicted box, even though it might have a higher IoU with the next predicted box.

Reply:
In this case I would say that the implementation you propose is just a different metric as the method to match the boxes is different. COCO (and I am pretty sure also PascalVOC) use the implementation shown in this PR, and I think it is relevant to provide a metric which is consistent with widely used ones. In the end it is a matter of "opinion" to say that the best matching pair is the one with highest IoU or the one with highest confidence.

However it is a nice addition to the PR. My opinion is that the mAP you propose should be an option in the pool rather than a replacement for the one implemented here. I guess most of the code could be shared between the two implementations.

Reply:
Sorry for returning to this old question. I also think that it's important that we use the same method as COCO and PASCAL VOC, so I wanted to check a third implementation to be confident that we're doing the right thing. Came across this implementation that says that it's computing COCO and PASCAL VOC mAP. [If the highest IoU is already matched, it continues to the next IoU.](https://github.com/bes-dev/mean_average_precision/blob/master/mean_average_precision/utils.py#L198) (Note that I'm not saying that the best matching pair is the one with the highest confidence, but the one with the highest IoU that's not matched to another prediction.) Just to check that we're talking about the same thing, could you point me to the code that you're referring to?

Reply:
I was mentioned in this PR a while ago and ever since I have been following this thread. Some colleagues and I have implemented COCO and PASCAL metrics in [this repository](https://github.com/rafaelpadilla/review_object_detection_metrics), which can be used with different file formats. The results with both COCO and PASCAL official implementations were verified.

The most time-consuming part was to implement the pytests, because it took me a while to run the MATLAB code with different file formats. Anyway, the tests are available for both COCO and PASCAL metrics and could help this repo to verify the results.

Reply:
some useful information: The boxes with the highest confidences have the highest priority on matching the ground-truths, that's why detections are ordered by confidence.

A detection is only matched with a groundtruth iff the IOU is above the IOU threshold (and the detection is set as a TP).
Once a ground-truth is matched with a detection, it cannot be matched with any other detection again (as mentioned above by @senarvi )

The difference between COCO and PASCAL is basically that COCO uses 101 points in the interpolation, and PASCAL uses all point for the interpolation. PASCAL calculates the AUC of the interpolated curve, while COCO takes the average of the interpolated curve by sampling the 101 points.

Hope it helps.

Reply:
@rafaelpadilla that doesn't answer the question what to do if the best ground truth is already matched. By looking at your code it seems to me that you [continue looping to see if any other ground truth is over the IoU threshold](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/src/evaluators/coco_evaluator.py#L322).

Reply:
I think this thread should be unresolved, since I pointed out that in those three other implementations, a FP is counted only if there are no unmatched targets with IoU over the threshold. In our implementation only the target with the best IoU is considered. @briankosw said that the standard mAP implementation compares only with the best match. I might be wrong, but I'd like to see a link to the standard implementation or the source of that information. Probably this won't make a big difference in the scores, so if you don't want to change this pull request anymore, you could at least add a note in the function documentation that this implementation only considers the target with the highest IoU.

Reply:
Hi @senavi, 

I have not been alerted about all updates in this thread, that's why I havent read your previous response.

In [our project](https://github.com/rafaelpadilla/review_object_detection_metrics/), the results were compared with [COCO](https://github.com/cocodataset/cocoapi/tree/master/PythonAPI/pycocotools)'s and [PASCAL](http://host.robots.ox.ac.uk/pascal/VOC/voc2012/#devkit)'s official implementations and they are identical. Our goal was to provide a tool that provides consistent results with all COCO and PASCAL AP variants accepting all types of bounding-box formats.

Differently than PASCAL, **COCO has a mAP variant that restricts the area of the bounding boxes to be considered**, namely  [_AP Across Scales_](https://cocodataset.org/#detection-eval).

Note that the code you mentioned, the groundtruths out of the area range are in the `gt_ignore` list. So, if a groundtruth is matched but is in the ignore list, it should not be considered. If that is the case, the loop continues to find other groundtruths. So, I think that answers your previous point. 

Yes, only the best match must be considered. Disconsidering the ignored boxes, if other detections overlap an already matched groundtruth, they should be taken as FP.

I dont know how the pytorch Lightning tasks are divided among the contributors, and as I am constantly busy with other activities, I preferred not to actively interfere in the code. But let me know if I can be useful to your project.

Best regards,
Rafael

Reply:
Sorry @rafaelpadilla, you're still not talking about the same thing. :)

Yes your code supports restricting the area of the bounding boxes. In this case, [on line 326 you stop if you have a match, and you encounter an ignored ground-truth](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/src/evaluators/coco_evaluator.py#L326). I'm not talking about that. We can assume that this feature is not used and `gt_ignore` is always `False`.

What I'm saying is that on [line 323, if you encounter a ground-truth that is already matched, you'll continue to the next ground truth](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/src/evaluators/coco_evaluator.py#L323). That is, you'll find the ground-truth with the highest IoU that hasn't been matched yet. You're not finding the highest IoU and failing if it has already been matched.

Reply:
It looks like your code is based on the `cocoeval.py` that you linked, since it's very similar. You can see that it also [continues to the next ground-truth if the first one is already matched](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/cocoeval.py#L280).

Reply:
@senarvi,

Yes, the code was based on the COCO's official implementation. In spite of that, sorry, but I think you did not get the point of how it is supposed to work.


You said: "you'll find the ground-truth with the highest IoU that hasn't been matched yet." then you continue: "You're not finding the highest IoU and failing if it has already been matched." It sounds contradictory. Why failing?

The expected behavior of the AP is to match the a groundtruth with at most one detection. The matching criteria is the IOU. I might be wrong, but I do not see a situation that the code will fail, as you mentioned. Please, give me an example illustrating when it will occur. 

Also note that line [324](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/src/evaluators/coco_evaluator.py#L324) must have the `continue` to avoid the current groundtruth (`g_idx`) being associated to the best match (Line [333](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/src/evaluators/coco_evaluator.py#L333)).

Reply:
@rafaelpadilla by "failing" I meant to say that you won't _stop_ if the target with the highest IoU has already been matched.

Yes, better to take a concrete example as you suggested. :) Say that we detect two similar bounding boxes, in more or less the same place. We also have two different targets, T1 and T2. Both detections have higher IoU with T1 than with T2, but all IoUs are above the threshold. Now we calculate the AP. We iterate the detections and match the first one with T1. What are we going to do with the next detection? Are we going to (i) find the target with the highest IoU (T1), notice that it's already matched, and count this as a FP, or (ii) find the target with the highest IoU that's not yet matched (T2), notice that the IoU is over the threshold, and match it with the second detection?

In this pull request we do matching using criterion (i), i.e. we [find the target with the highest IoU](https://github.com/PyTorchLightning/pytorch-lightning/blob/4b1e5a3e176e69fd7dcd9406728e8c765c8d94a0/pytorch_lightning/metrics/functional/object_detection.py#L63) and if it's already matched, that's it. Most other implementations use criterion (ii), including the official COCO implementation. The COCO implementation would notice that the first target is matched, [continue](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/src/evaluators/coco_evaluator.py#L324) to the next target, notice that it's not matched yet and match it.

[This implementation](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/Evaluator.py#L111) uses criterion (i). For both detections it would find the target with the highest IoU (T1). It would match T1 with the first detection, and for the second detection it would see that T1 is already matched and count it as a FP. I downloaded the VOC development kit and it seems that it also uses criterion (i). So I guess this is a difference between PASCAL and COCO,

Reply:
@senarvi,

With the practical example that you provided, we bring another difference between COCO and PASCAL implementations. Yes, you are correct on your conclusions. 

If your implementation uses the **approach (i), you are following PASCAL criterion**. As you could see in [our PASCAL implementation](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/src/evaluators/pascal_voc_evaluator.py) and confirmed in their toolkit file `VOCevaldet.m`.

COCO's implementation uses the approach **(ii), where it will continue looping through the other non-matched groundtruths to find a possible match**.

But our code does not fail, as it is not considered a failure. It is just a different implementation approach used by COCO and PASCAL. Our repository implements both COCO ([coco_evaluator.py](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/src/evaluators/coco_evaluator.py)) and PASCAL ([pascal_voc_evaluator.py](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/src/evaluators/pascal_voc_evaluator.py)) approaches, [so the user can pick one](https://github.com/rafaelpadilla/review_object_detection_metrics/blob/main/data/images/printshot_main_screen.png). 

As I have seen on the discussions, you are still deciding if both will be implemented, right? I think it should be very clear to the user which one you are offering. This will avoid confusions if users publish their results saying he used COCO's metric, when in fact it was PASCAL's.

Reply:
That's right, your code doesn't fail. I never meant to say that your code or any other code is failing. I simply wanted to verify whether a standard implementation would also stop (that is, fail to find a match), if the target with the highest IoU is matched already.

Anyway, since this PR should be merged soon, it's better to just add a mention in the documentation that this function follows the PASCAL style matching of detections and targets (considers only the target with the highest IoU), while COCO finds a match if any target that has IoU higher than the threshold is not yet matched. Right @briankosw?

Reply:
Yes. No worries. I didnt take it badly :)